File: NeetCode/Intervals/Meeting_Rooms/meeting__rooms.py
from typing import Optional, List


class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end


class Solution:
    def canAttendMeetings(self, intervals: List[Interval]) -> bool:
        intervals.sort(key=lambda x: x.start)

        for index in range(len(intervals) - 1):
            if intervals[index].end > intervals[index + 1].start:
                return False

        return True


if __name__ == "__main__":
    sol = Solution()
    intervals = [(0, 30), (5, 10), (15, 20)]
    print(sol.canAttendMeetings(intervals))
    print("Running Solution...")

File: NeetCode/Intervals/Meeting_Rooms/Meeting_Rooms.cpp
// Meeting_Rooms.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;
class Interval
{
public:
    int start, end;
    Interval(int start, int end)
    {
        this->start = start;
        this->end = end;
    }
};

class Solution
{
public:
    bool canAttendMeetings(vector<Interval> &intervals)
    {
        sort(intervals.begin(), intervals.end(),
             [](const Interval &a, const Interval &b)
             {
                 return a.start < b.start;
             });
        for (int index = 0; index < intervals.size() - 1; index++)
        {
            if (intervals[index].end > intervals[index + 1].start)
            {
                return false;
            }
        }
        return true;
    }
};

/*
leetcode is different because there they use vector<vector<int>> instead of interval

class Solution {
public:
    bool canAttendMeetings(vector<vector<int>>& intervals) {
        if(intervals.size()==0){
            return true;
        }
        sort(intervals.begin(), intervals.end(),
     [](const vector<int>& a, const vector<int>& b) {
         return a[0] < b[0];
     });
        for (int index = 0; index < intervals.size() - 1; index++)
        {
            if(intervals[index][1]> intervals[index+1][0]){
                return false;
            }
        }
        return true;
    }
};
*/
